Switch to adjacent buffer with alt+arrow up/down

20after4 opened the pull request:
A rough implementation of #511

probably only works in nested bufferlist.

lorenzhs commented on js/glowingbear.js:
```diff
@@ -727,6 +727,26 @@ weechat.controller('WeechatCtrl', ['$rootScope', '$scope', '$store', '$timeout',
 
+    $rootScope.switchToAdjacentBuffer = function(direction) {
+        // direction is +1 for next buffer, -1 for previous buffer
+        var sortedBuffers = _.sortBy($scope.getBuffers(), 'fullName');
```
this should probably mimic what the angular stuff does in the html and use `$rootScope.predicate` for sorting instead of hardcoded `'fullName'`

lorenzhs commented on js/glowingbear.js:
```diff
@@ -727,6 +727,26 @@ weechat.controller('WeechatCtrl', ['$rootScope', '$scope', '$store', '$timeout',
+        var bufferKeys = Object.keys(sortedBuffers);
+
+        // Try to find adjacent
+        for (i in sortedBuffers) {
```
you could do something like this:
```js
var index = sortedBuffers.indexOf(activeBuffer);
if (index >= 0) {
    var newBuffer = sortedBuffers[index + direction];
    if (newBuffer) {
        $scope.setActiveBuffer(newBuffer.id);
    }
}
```

lorenzhs commented on js/glowingbear.js:
```diff
@@ -727,6 +727,26 @@ weechat.controller('WeechatCtrl', ['$rootScope', '$scope', '$store', '$timeout',
+        for (i in sortedBuffers) {
+            buffer = sortedBuffers[i];
+            if (buffer === activeBuffer) {
+                var newKey = parseInt(i)+parseInt(direction);
```
why do you need the `parseInt` here? shouldn't it already be a number?

lorenzhs commented on js/inputbar.js:
```diff
@@ -294,6 +294,14 @@ weechat.directive('inputBar', function() {
+                if ($event.altKey && (code === 38 || code === 40)) {
+                    $event.preventDefault();
+                    var direction = code - 39;
+                    console.log('direction:', direction)
```
please remove the debug logging. I use it a lot myself during development but we should keep it clean in production code (maybe we could use $log with appropriate levels but that's another discussion)

lorenzhs commented on js/inputbar.js:
```diff
@@ -294,6 +294,14 @@ weechat.directive('inputBar', function() {
+                    $event.preventDefault();
+                    var direction = code - 39;
+                    console.log('direction:', direction)
+                    $rootScope.switchToAdjacentBuffer(direction)
```
missing semicolon

lorenzhs commented:
Thanks for implementing this! I hope my comments make sense and that I didn't make too many mistakes typing out code in a comment box ;) Oh and please squash the commits in the end so we don't clobber up the history.

20after4 commented:
@lorenzhs: thanks for all the help! I squashed the commits and addressed your feedback. And it seems to work nicely for me in nested bufferlist mode as well as with a flat list.

lorenzhs commented:
Great! Some of the imgur bits seem to have gotten caught up in your last commit though, I'm not sure what exactly went wrong there but we'd like to keep them ;)

lorenzhs commented on js/glowingbear.js:
```diff
@@ -727,6 +727,19 @@ weechat.controller('WeechatCtrl', ['$rootScope', '$scope', '$store', '$timeout',
+    $rootScope.switchToAdjacentBuffer = function(direction) {
+        // direction is +1 for next buffer, -1 for previous buffer
+        var sortedBuffers = _.sortBy($scope.getBuffers(), $rootScope.predicate);
+        var i, buffer;
```
those two are now unused

20after4 commented:
Ok that should be ready to land then?

torhve commented on js/inputbar.js:
```diff
@@ -294,6 +294,13 @@ weechat.directive('inputBar', function() {
                     return true;
                 }
 
+                if ($event.altKey && (code === 38 || code === 40)) {
```
I would prefer it if this code also checked that ctrlkey wasn't down. As right now this would also work with ctrl-alt-arrow. In the future we might would want to do something different for that, or this might trigger something else in some other layout that we don't know about. And in general, it's better to be explicit if you mean alt+arrow and *only* alt-arrow, not all the other combinations.

lorenzhs replied:
Good catch! The current state interferes with workspace switching for me. Please check that Ctrl isn't pressed a swell.

20after4 replied:
done

lorenzhs commented:
LGTM, haven't tested it though

lorenzhs commented:
has a weird bug for me, it goes from `#weechat-offtopic` to `NickServ` to `dcormier` which isn't particularly alphabetical. Maybe the uppercase character is messing with the sort order (because it's a lower ASCII char). I suspect angular's sorting ignores case while underscore's doesnt?

20after4 commented:
That is an odd bug. I don't see it on my end even though I do have some uppercase buffers. hmm

20after4 commented:
@lorenzhs: try it now?

lorenzhs commented:
awesome, it works now. Thanks a lot!

fuzzy76 commented:
This should probably have been mentioned in the landing page as well.

lorenzhs commented:
Good idea! I just proposed #749
